src/codegen/compile.ml

```diff
@@ -7129,29 +6899,30 @@ module MakeSerialization (Strm : Stream) = struct
 
   (* This value is returned by deserialize_go if deserialization fails in a way
      that should be recoverable by opt parsing.
-     By virtue of being a deduped static value, it can be detected by pointer
-     comparison.
+     It is an (invalid) sentinel pointer value (in skewed format) and can be used for pointer comparison.
+     It will be never placed on the heap and must not be dereferenced.
+     If unskewed, it refers to the unallocated last Wasm memory page.
   *)
-  let coercion_error_value env : int32 =
-    Tagged.shared_static_obj env Tagged.CoercionFailure []
+  let coercion_error_value env = 0xffff_fffdl
 
   (* See Note [Candid subtype checks] *)
   let with_rel_buf_opt env extended get_typtbl_size1 f =
     if extended then
       f (compile_unboxed_const 0l)
     else
-      get_typtbl_size1 ^^ get_typtbl_size env ^^
+      get_typtbl_size1 ^^ count_type_offsets env ^^
       E.call_import env "rts" "idl_sub_buf_words" ^^
       Stack.dynamic_with_words env "rel_buf" (fun get_ptr ->
-        get_ptr ^^ get_typtbl_size1 ^^ get_typtbl_size env ^^
+        get_ptr ^^ get_typtbl_size1 ^^ count_type_offsets env ^^
         E.call_import env "rts" "idl_sub_buf_init" ^^
         f get_ptr)
 
   (* See Note [Candid subtype checks] *)
   let idl_sub env t2 =
```

Comment:
This looks pretty expensive to me now since every idl_sub check (which could occur in a loop) will allocate several blobs for the global type table. It might be better to store the data once in a reserved area or at the base of the (Rust) stack and re-use it for every subtype check. Will need to think about this.

Reply:
Ok, I think the way to solve this is to add some registers to module Registers and only load the global type table fields, ideally to the stack, once on entry to `deserialize_go` (where all the other registers are set). That way we only allocate one or two blobs for each call to deserialize_go, not one per call to idl_sub. Though I guess that does mean we allocate even when there is no call to idl_sub....

Reply:
Thank you very much for having noticed this performance bug! Also thanks for the performance analysis. I will fix this.

Reply:
Thank you very much for the performance analysis (https://github.com/dfinity/motoko/pull/4481). I believe the problem is in `idl_sub` where the Blob is unnecessarily reloaded. I missed to optimize that case (pooling).

Here, `count_type_offsets` sounds worse than it is: It only reads a global variable, the number of if type offsets.

Reply:
Excellent, glad it was easy to fix. There might be further optimization opportunities to avoid recalculating the absolute indices but maybe not worth it.

Reply:
Thanks that you noticed there is an additional dynamic blob allocation inm `idl_sub` in the runtime system to map relative offsets to static pointer table. This could also be recycled across repeated `idl_sub` calls of the same deserialization procedure (staying within the bounds of the same IC message, no GC increment running in between).

Reply:
Thanks a lot for preparing https://github.com/dfinity/motoko/pull/4497. This will heal the regression.